File: 0473-matchsticks-to-square/0473-matchsticks-to-square.js
/**
 * @param {number[]} matchsticks
 * @return {boolean}
 */
var makesquare = function (matchsticks) {
    let total = 0;
    let maxLength = 0;

    for (const stick of matchsticks) {
        total += stick;
        maxLength = Math.max(maxLength, stick);
    }

    const targetSideLength = Math.floor(total / 4);
    const remainder = total % 4;

    if (remainder !== 0 || targetSideLength < maxLength) {
        return false;
    }

    matchsticks.sort((a, b) => b - a);

    const sticks = new Array(4).fill(0);

    const dfs = function (index, sticks) {
        if (index === matchsticks.length) {
            return true;
        }

        for (let stickIndex = 0; stickIndex < 4; stickIndex++) {

            if (stickIndex > 0 && sticks[stickIndex] === sticks[stickIndex - 1]) continue;

            sticks[stickIndex] += matchsticks[index];

            if (sticks[stickIndex] <= targetSideLength && dfs(index + 1, sticks)) {
                return true;
            }

            sticks[stickIndex] -= matchsticks[index];
        }

        return false;
    }

    return dfs(0, sticks);
};

File: 0473-matchsticks-to-square/0473-matchsticks-to-square.py
class Solution:
    def makesquare(self, matchsticks: List[int]) -> bool:
        total = sum(matchsticks)
        maxLength = max(matchsticks)

        # remainder = total % 4
        # targetMaxStick = total // 4

        targetMaxStick, remainder = divmod(total, 4)

        if remainder != 0 or targetMaxStick < maxLength:
            return False

        matchsticks.sort(reverse=True)

        sticks = [0] * 4

        def dfs(index, sticks):
            if index == len(matchsticks):
                return True

            for i in range(4):
                if i > 0 and sticks[i] == sticks[i - 1]:
                    continue

                sticks[i] += matchsticks[index]

                if sticks[i] <= targetMaxStick and dfs(index + 1, sticks):
                    return True

                sticks[i] -= matchsticks[index]

            return False

        return dfs(0, sticks)
